### PR TITLE
Fixing the materializer registry to spot sub-classes of defined types

### PR DESCRIPTION
**Pull request by bcdurak**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Small adjustment to the materializer registry to recognize if the provided Input/Output annotation points to a subclass of a previously defined type.
Special case: If we are handling the specific case, where the given type inherits from more than one class, it checks whether the multiple base classes map to at least one (if more the same materializer) and uses it only if the condition is cleared.

### CODE REVIEW

**Review by AlexejPenner — APPROVED** [review]
Love it, especially the catch in case the object at hand is subclassing multiple classes that have differing materializers.

File: src/zenml/materializers/default_materializer_registry.py [hunk 1]
```diff
@@ -71,10 +71,24 @@ def __getitem__(self, key: Type[Any]) -> Type["BaseMaterializer"]:
...
+            compatible_subclasses = {
+                self.materializer_types[t]
+                for t in self.materializer_types
+                if issubclass(key, t)
+            }
+            if len(compatible_subclasses) == 1:
+                return self.materializer_types[compatible_subclasses[0]]
+            else:
+                raise KeyError(
+                    f"Type {key} is subclassing more than one type, thus it "
```
Comment (schustmi) [inline src/zenml/materializers/default_materializer_registry.py:83]: Maybe we can also include the `compatible_subclasses` in the error somehow to show which materializers overlap?

**Review by schustmi — CHANGES_REQUESTED** [review]
Really nice solution with the set to check if many subclasses are handled by one or multiple materializers!
Just have one small comment on the log message, and I think it would also be nice to have some tests to make sure we don't break this in the future?

**Comment by bcdurak** [comment]
@schustmi Great idea! I will add the tests right away!

**Comment by bcdurak** [comment]
@schustmi I have added the tests!

File: src/zenml/artifacts/type_registry.py [hunk 2]
```diff
@@ -44,7 +44,15 @@ def register_integration(
 
     def get_artifact_type(self, key: Type[Any]) -> List[Type["BaseArtifact"]]:
         """Method to extract the list of artifact types given the data type"""
-        return self._artifact_types[key]
+        if key in self._artifact_types:
```
Comment (schustmi) [inline src/zenml/artifacts/type_registry.py:47]: Is there a possibility that no artifact types are registered for some class? If yes I feel like this method should raise a `KeyError` instead of whatever happens when calling `pop()` on an empty list. (And the KeyError should probably be in the docstring)

File: tests/unit/materializers/test_base_materializer.py [hunk 3]
```diff
@@ -11,3 +11,207 @@
...
+
+import pytest
+
+from zenml.exceptions import StepInterfaceError
+from zenml.materializers import BuiltInMaterializer
+from zenml.materializers.base_materializer import BaseMaterializer
+from zenml.steps import Output, step
+
+
+def test_call_step_with_missing_materializer_for_type():
```
Comment (schustmi) [inline tests/unit/materializers/test_base_materializer.py:24]: More of a philosophical question, but those existing tests are testing code that is inside `base_step.py` and not actually inside the materializer. (The actual base materializer class is essentially empty). Should we place the tests in the file corresponding to the code they test (would be `base_step.py` and `default_materializer_registry.py` in this case) or group them thematically like you did here?

**Review by schustmi — CHANGES_REQUESTED** [review]
Really love the new tests, just two small questions

File: src/zenml/artifacts/type_registry.py (hunk 2 above)
Comment (bcdurak) [reply src/zenml/artifacts/type_registry.py:47]: Given the current state of the implementation, when the execution reaches that point there is no way for that list to be empty. It is due to the fact that the entries of the `type_registry` mirror the entries of the `materializer_registry` in terms of keys and the `materializer_registry` gets called first. So if there are no compatible subclasses, `materializer_registry` will fail before the code reaches that point. However, the implementation may change in the future and I agree that it would make sense to cover the edge-cases as we build it up. I will create a short fix for it!

File: tests/unit/materializers/test_base_materializer.py (hunk 3 above)
Comment (bcdurak) [reply tests/unit/materializers/test_base_materializer.py:24]: I was a bit hesitant about this as well. I felt like we were testing the connection between the steps and the materializers. But on second thought, I think you are right. We are testing the code within the `base_step.py` and the tests should be placed accordingly. I am on it!

Comment (bcdurak) [reply tests/unit/materializers/test_base_materializer.py]: Changed the structure of the tests back.

File: src/zenml/materializers/default_materializer_registry.py [hunk 4]
```diff
@@ -71,10 +72,25 @@ def __getitem__(self, key: Type[Any]) -> Type["BaseMaterializer"]:
...
-                f"Type {key} does not have a default `Materializer`! Please "
-                f"specify your own `Materializer`."
-            )
+            compatible_subclasses = {
+                self.materializer_types[t]
+                for t in self.materializer_types
+                if issubclass(key, t)
+            }
+            if len(compatible_subclasses) == 1:
+                return compatible_subclasses.pop()
```
Comment (jwwwb) [inline src/zenml/materializers/default_materializer_registry.py:81]: Any reason to use `.pop()` instead of `[0]` here? it's mutating the list in place, but it's not like that is ever reused again anyway, so it should be garbage collected.

File: src/zenml/materializers/default_materializer_registry.py [hunk 5]
```diff
@@ -84,7 +100,7 @@ def get_materializer_types(
 
     def is_registered(self, key: Type[Any]) -> bool:
         """Returns if a materializer class is registered for the given type."""
-        return key in self.materializer_types
+        return any([issubclass(key, t) for t in self.materializer_types])
```
Comment (jwwwb) [inline src/zenml/materializers/default_materializer_registry.py:103]: `any()` will also accept a generator instead of a list, so you can simply remove the square brackets and it will still work (and be more efficient).

File: src/zenml/artifacts/type_registry.py [hunk 6]
```diff
@@ -44,7 +44,15 @@ def register_integration(
 
     def get_artifact_type(self, key: Type[Any]) -> List[Type["BaseArtifact"]]:
```
Comment (jwwwb) [inline src/zenml/artifacts/type_registry.py:45]: at a glance, doesn't this pull request contain two functions that do more ore less the same thing? check a dictionary for the presence of a key, return the associated value to that if it exists and otherwise search for supertypes, returning that associated value only if there is exactly one, else raising an error? seems like that could be in a generic function to make sure if something changes in the future, it has to be changed in one place only.

**Review by jwwwb — CHANGES_REQUESTED** [review]

File: src/zenml/artifacts/type_registry.py (hunk 2 above)
Comment (bcdurak) [reply src/zenml/artifacts/type_registry.py:60]: Added the checks to the `type_registry` as well and updated the docstrings.

File: src/zenml/materializers/default_materializer_registry.py [hunk 7]
```diff
@@ -71,10 +72,25 @@ def __getitem__(self, key: Type[Any]) -> Type["BaseMaterializer"]:
         if key in self.materializer_types:
             return self.materializer_types[key]
         else:
-            raise KeyError(
-                f"Type {key} does not have a default `Materializer`! Please "
-                f"specify your own `Materializer`."
-            )
+            compatible_subclasses = {
```
[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/materializers/default_materializer_registry.py]: Thanks for the tip regarding the naming, I have already changed it accordingly. 

As for the set, I am using it as a safety net. In the future, we might be dealing with classes that inherit from more than one base class. In this case, we need to figure out which `materializer` to use if it is not explicitly defined. Let's go over the example where the annotation says that the output will be an instance of class `C` that inherits from class `A` and `B`. We first check whether `C` is registered and use it if available. If not we look for `A` or `B`. If only one is registered and the other isn't, I will be using the `materializer` associated with the superclass which is already registered. In the case where they are both registered, I would need to compare them. That's where the `set` comes into play. If both `A` and `B` point to the same `materializer`, the `compatible_superclasses` will end up being 1 and I can safely assume that I can use that materializer. If they point to different `materializers`, I will raise an error and ask our users to explicitly define it.

File: src/zenml/materializers/default_materializer_registry.py (hunk 5 above)
Comment (bcdurak) [reply src/zenml/materializers/default_materializer_registry.py]: Nice catch! Changed it accordingly.

File: src/zenml/artifacts/type_registry.py [hunk 8]
```diff
@@ -44,7 +44,15 @@ def register_integration(
...
-        return self._artifact_types[key]
+        if key in self._artifact_types:
+            return self._artifact_types[key]
+        else:
+            compatible_subclasses = [
+                self._artifact_types[t]
+                for t in self._artifact_types
+                if issubclass(key, t)
+            ]
+            return compatible_subclasses.pop()
```
[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/artifacts/type_registry.py]: There is no defined order to the list. However, in order for the algorithm to kick in for subclassing keys, I need to iterate over the whole dictionary and figure out if they fulfill certain criteria.

File: src/zenml/materializers/default_materializer_registry.py (hunk 4 above)
Comment (bcdurak) [reply src/zenml/materializers/default_materializer_registry.py]: `compatible_superclass` here is a set and therefore it is not a subscriptable object. That's why decided to use the `pop` method. Not sure how to improve the garbage collection though.

File: src/zenml/artifacts/type_registry.py (hunk 6 above)
Comment (bcdurak) [reply src/zenml/artifacts/type_registry.py:46]: Good idea to implement a generic function. Right now, there is a hiccup in our implementation. If the key is not present but its supertypes are, we need to check whether the corresponding values are the same as well. In our `default_materializer_registry`, the values are actual `Materializer`s which we can compare between each other. However, in the `type_registry`, the values are `list`s so they are unhashable and incomparable. That's why the implementation differs just a little bit between them. I have an idea on how to fix this and will update this thread as it progresses.

**Comment by bcdurak** [comment]
@jwwwb following our discussion from last Friday, I have made some changes to the PR. Thank you for the great tips. I have also included a fix converting `ASSOCIATED_TYPES` and `ASSOCIATED_ARTIFACT_TYPES` from lists to tuples within the `BaseMaterializer` for the sake of immutability.

File: src/zenml/artifacts/type_registry.py (hunk 6 above)
Comment (bcdurak) [reply src/zenml/artifacts/type_registry.py]: Updated the main thread!

File: src/zenml/artifacts/type_registry.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/artifacts/type_registry.py:65]: Looks awesome now!

**Review by jwwwb — APPROVED** [review]

**Review by schustmi — APPROVED** [review]
Really nice!
